### ISSUE TITLE
Back `MerkleTree::data` with `mmap`

### ISSUE THREAD

**Issue by schomatis** [filecoin-project/merkle_light#8]
The complication is that the `memmap::MmapMut` solution we normally use only covers the `u8` type whereas `MerkleTree::data` is a `Vec<T>` where `T: Ord + Clone + AsRef<[u8]> + Sync + Send`. See https://github.com/filecoin-project/merkle_light/pull/7 as an example of a failed solution.

As suggested by dig these are the current steps to investigate a more elaborate implementation:

> because `T` has `AsRef<[u8]>` you need to rely on that
you need to change the method to require two things
a) having the data amount known before hand
b) having the size of `T` in bytes known
then you can create an mmap large enough to hold the things
get it as a `&mut [u8]`
and then use https://docs.rs/rayon/1.0.2/rayon/slice/trait.ParallelSliceMut.html#method.par_chunks_mut to write into each chunk the hash

**Comment by dignifiedquire** [comment]
You will also need to change the `data` field itself to be an mmap thing, rather than a `vec` otherwise you will not gain any benefits.

**Comment by schomatis** [comment]
> You will also need to change the `data` field of the Tree itself to be an mmap thing, rather than a `vec` otherwise you will not gain any benefits.

The benefit I want in this case is to have `Vec` allocated through `mmap` instead of `malloc`, I was assuming (but actually didn't check the code) that conversions like `to_vec()` passed ownership of the slice (allocated with `mmap`) to the underlying `RawVec` in which case we could keep using `Vec` (knowing where it came from), is this assumption incorrect?

**Comment by dignifiedquire** [comment]
looking at the code, you need to use `into_vec` if you want to do that. `to_vec` copies and allocates on the heap.

**Comment by dignifiedquire** [comment]
but I think that is not available on mmemap

**Comment by schomatis** [comment]
@dignifiedquire Rereading your proposed design, I'd like to approach this in an iterative fashion, my first goal would be to get it to build without errors (even if it doesn't allocate the correct space for the tree or if it doesn't actually compute the hashes and it fails at runtime), so I'm supposing at the moment I have (a) and (b), my question then would be: I'm at the point where `mmap` is a slice of `&mut [u8]`, at which point can I (and how) convert that to a `Vec<T>`? is that what `par_chunks_mut` would help me with? (if that's indeed what I should do, I may be misinterpreting your proposal)

**Comment by dignifiedquire** [comment]
Sorry for being confusing, tldr mmap and Vec are either or. Now that is not a big issue because this library relies on slices rather than the fact that the backing storage is a vector.

**Comment by schomatis** [comment]
> tldr mmap and Vec are either or.

Ok, so the first step would then be to change `MerkleTree::data` from a `Vec<T>` to `&[u8]` (slice), removing the `T` parameter type. Am I following this correctly?

**Comment by schomatis** [comment]
So, I think the missing piece of the puzzle here is that I need to do an `unsafe` call to `mem::transmute` to change `[u8]` for `[T]` (adjusting sizes as appropriate), removing then

https://github.com/filecoin-project/merkle_light/blob/63a609decf3cc552b0aa79477b828f4a493f725d/merkle/src/lib.rs#L155

Unless told otherwise, I'm going to go with that for the time being, at least this let's me compile the code..

**Comment by dignifiedquire** [comment]
Note `T` is not actually `u8`, it is `[u8]` (with differing sizes), so this will make a very wrong thing and not work as expected, when simply doing a transmute.

**Comment by schomatis** [comment]
> Note `T` is not actually `u8`, it is `[u8]` (with differing sizes), so this will make a very wrong thing and not work as expected, when simply doing a transmute.

See https://github.com/filecoin-project/merkle_light/pull/12/files#diff-51a9ffca05030231847ab34ac3206a3aR220.
